File: api/grpc/src/services/layouts.rs
use grpc::{ServerRequestSingle, ServerResponseUnarySink};

use mizer_api::handlers::LayoutsHandler;
use mizer_api::models::*;

use crate::protos::*;
use mizer_api::RuntimeApi;

impl<R: RuntimeApi> LayoutsApi for LayoutsHandler<R> {
    fn get_layouts(
        &self,
        _: ServerRequestSingle<GetLayoutsRequest>,
        resp: ServerResponseUnarySink<Layouts>,
    ) -> grpc::Result<()> {
        let layouts = self.get_layouts();

        resp.finish(layouts)
    }

    fn add_layout(
        &self,
        req: ServerRequestSingle<AddLayoutRequest>,
        resp: ServerResponseUnarySink<Layouts>,
    ) -> grpc::Result<()> {
        let layouts = self.add_layout(req.message.name);

        resp.finish(layouts)
    }

    fn remove_layout(
        &self,
        req: ServerRequestSingle<RemoveLayoutRequest>,
        resp: ServerResponseUnarySink<Layouts>,
    ) -> grpc::Result<()> {
        let layouts = self.remove_layout(req.message.id);

        resp.finish(layouts)
    }

    fn rename_layout(
        &self,
        req: ServerRequestSingle<RenameLayoutRequest>,
        resp: ServerResponseUnarySink<Layouts>,
    ) -> grpc::Result<()> {
        let layouts = self.rename_layout(req.message.id, req.message.name);

        resp.finish(layouts)
    }

    fn rename_control(
        &self,
        _: ServerRequestSingle<RenameControlRequest>,
        _: ServerResponseUnarySink<LayoutResponse>,
    ) -> grpc::Result<()> {
        todo!()
    }

    fn move_control(
        &self,
        _: ServerRequestSingle<MoveControlRequest>,
        _: ServerResponseUnarySink<LayoutResponse>,
    ) -> grpc::Result<()> {
        todo!()
    }

    fn update_control(&self, req: ServerRequestSingle<UpdateControlRequest>, resp: ServerResponseUnarySink<LayoutResponse>) -> grpc::Result<()> {
        self.update_control(req.message.layout_id, req.message.control_id, req.message.decorations.unwrap());

        resp.finish(Default::default())
    }

    fn remove_control(
        &self,
        req: ServerRequestSingle<RemoveControlRequest>,
        resp: ServerResponseUnarySink<LayoutResponse>,
    ) -> grpc::Result<()> {
        self.remove_control(req.message.layout_id, req.message.control_id);

        resp.finish(Default::default())
    }

    fn add_control(&self, req: ServerRequestSingle<AddControlRequest>, resp: ServerResponseUnarySink<LayoutResponse>) -> grpc::Result<()> {
        self.add_control(req.message.layout_id, req.message.node_type, req.message.position.unwrap()).unwrap();

        resp.finish(Default::default())
    }

    fn add_existing_control(&self, req: ServerRequestSingle<AddExistingControlRequest>, resp: ServerResponseUnarySink<LayoutResponse>) -> grpc::Result<()> {
        self.add_control_for_node(req.message.layout_id, req.message.node.into(), req.message.position.unwrap()).unwrap();

        resp.finish(Default::default())
    }
}

File: ui/lib/api/grpc/layouts.dart
import 'package:grpc/grpc.dart';
import 'package:mizer/api/contracts/layouts.dart';
import 'package:mizer/protos/layouts.pbgrpc.dart';
import 'package:mizer/protos/nodes.pbenum.dart';

class LayoutsGrpcApi implements LayoutsApi {
  final LayoutsApiClient client;

  LayoutsGrpcApi(ClientChannel channel) : this.client = LayoutsApiClient(channel);

  @override
  Future<Layouts> getLayouts() {
    return this.client.getLayouts(GetLayoutsRequest());
  }

  @override
  Future<Layouts> addLayout(String name) {
    return this.client.addLayout(AddLayoutRequest(name: name));
  }

  @override
  Future<Layouts> removeLayout(String id) {
    return this.client.removeLayout(RemoveLayoutRequest(id: id));
  }

  @override
  Future<Layouts> renameLayout(String id, String name) {
    return this.client.renameLayout(RenameLayoutRequest(id: id, name: name));
  }

  @override
  Future<void> deleteControl(String layoutId, String id) async {
    var request = RemoveControlRequest(controlId: id, layoutId: layoutId);
    await this.client.removeControl(request);
  }

  @override
  Future<void> moveControl(String layoutId, String id, ControlPosition position) async {
    var request = MoveControlRequest(layoutId: layoutId, controlId: id, position: position);
    await this.client.moveControl(request);
  }

  @override
  Future<void> renameControl(String layoutId, String id, String name) async {
    var request = RenameControlRequest(layoutId: layoutId, controlId: id, name: name);
    await this.client.renameControl(request);
  }

  @override
  Future<void> addControl(String layoutId, Node_NodeType nodeType, ControlPosition position) async {
    var request = AddControlRequest(layoutId: layoutId, nodeType: nodeType, position: position);
    await this.client.addControl(request);
  }

  @override
  Future<void> addControlForNode(String layoutId, String nodeId, ControlPosition position) async {
    var request = AddExistingControlRequest(layoutId: layoutId, node: nodeId, position: position);
    await this.client.addExistingControl(request);
  }

  @override
  Future<void> updateControl(String layoutId, String id, ControlDecorations decoration) async {
    var request = UpdateControlRequest(layoutId: layoutId, controlId: id, decorations: decoration);
    await this.client.updateControl(request);
  }
}
